sky/data/storage.py

```diff
@@ -458,7 +489,8 @@ def __init__(self,
                  stores: Optional[Dict[StoreType, AbstractStore]] = None,
                  persistent: Optional[bool] = True,
                  mode: StorageMode = StorageMode.MOUNT,
-                 sync_on_reconstruction: bool = True) -> None:
+                 sync_on_reconstruction: bool = True,
+                 bucket_sub_path: Optional[str] = None) -> None:
```

Comment:
Will handle `mount` and `delete`

>we should delete the bucket if and only there are no other sub dirs present in the bucket
Is it like a mechanism that stores the subdirectory mapping of a bucket in the global user state and triggers bucket deletion if the state is cleared?

Reply:
Thanks for your reply @romilbhardwaj 

For mount:
* The mount supports both COPY and MOUNT modes under `sub_dir` in the current code.

For delete:

* The [delete  storage](https://github.com/skypilot-org/skypilot/blob/6fda9fd5717cfb08afd5ad7e8cff0a8f95bf6200/sky/data/storage.py#L940-L982) code you pointed out to me is triggered by manually executing `sky storage delete`. I think we should keep the way `sky storage delete` deletes a certain bucket, instead of deleting the` sub_dir` of a bucket, even if it's a `sub_dir` mount. I suggest we don't change the behavior of the storage delete function.

* The [_teardown_ephemeral_storage](https://github.com/skypilot-org/skypilot/blob/6fda9fd5717cfb08afd5ad7e8cff0a8f95bf6200/sky/backends/cloud_vm_ray_backend.py#L3506-L3511) code you mentioned earlier—I can't trigger this function. When I execute `sky jobs launch ~/hello-sky/work_dir_1.yaml` and, after the job succeeds, execute `sky down sky-jobs-controller-70ac5fe9`, the storage still exists; the function is not triggered. Do you know how to trigger this function so that I can test the behavior?
    * If we don't want to change the behaviour of  `sky storage delete`, should we change the behaviour of `_teardown_ephemeral_storage`?

Reply:
For mount, I don't think this PR has the desired behavior. I think we want the `bucket_sub_path` to be invisible to the application which will use the mounted path, similar to [subPaths in Kubernetes volumes](https://kubernetes.io/docs/concepts/storage/volumes/#using-subpath). I.e., the mounted path should contain the contents the user uploaded, not the root of the bucket.

For example, this task:
```
import sky

resources = sky.Resources(cloud=sky.clouds.AWS())
t = sky.Task(run='ls -l /mydir')
t = t.set_resources(resources)
s = sky.Storage(name='test-bucket-romilb', source='~/tmp-workdir', bucket_sub_path='mypath', mode=sky.StorageMode.MOUNT)
t = t.set_storage_mounts({'/mydir': s})
print(t.storage_mounts)
sky.launch(t, cluster_name='test')
```

Task output in this branch:
```
(task, pid=2062) total 4
(task, pid=2062) drwxr-xr-x 2 ubuntu ubuntu 4096 Nov  9 21:40 mypath
# This should have been the contents of my tmp-workdir
```

Reply:
I see now

Reply:
Thanks @romilbhardwaj 

This code works now under latest commit:

```python
import sky

resources = sky.Resources(cloud=sky.clouds.AWS())
t = sky.Task(run='sleep 10; ls -l /storage_mount')
t = t.set_resources(resources)
s = sky.Storage(name='test-bucket-zpoint', source='~/Desktop/dir1', bucket_sub_path='mypath_script', mode=sky.StorageMode.MOUNT)
s.add_store(sky.StoreType.GCS)
t = t.set_storage_mounts({'/storage_mount': s})
# storage mount won't be triggered unless file mounts are set
t = t.set_file_mounts({'/file_mount': '~/Desktop/dir1'})
print(t.file_mounts)
print(t.storage_mounts)
sky.launch(t, cluster_name='test-dir')
```

Reply:
Awesome, thanks! Mount mode works nicely now. Can we also make sub-dir work for COPY mode?

E.g., this task:
```
import sky

resources = sky.Resources(cloud=sky.clouds.AWS())
#t = sky.Task(run='ls -lH /s1 /s2 /s3')
t = sky.Task(run='ls -lH /s1 /s3')
t = t.set_resources(resources)
mode=sky.StorageMode.COPY
s1 = sky.Storage(name='test-bucket-romilb1', source='~/tmp-workdir', bucket_sub_path='mypath', mode=mode)
s1.add_store('S3')
s2 = sky.Storage(name='test-bucket-romilb2', source='~/tmp-workdir', bucket_sub_path='mypath', mode=mode)
s2.add_store('GCS')
s3 = sky.Storage(name='test-bucket-romilb3', source='~/tmp-workdir', bucket_sub_path='mypath', mode=mode)
s3.add_store('AZURE')
# t = t.set_storage_mounts({'/s1': s1, '/s2': s2, '/s3': s3})
t = t.set_storage_mounts({'/s1': s1, '/s3': s3})
print(t.storage_mounts)
sky.launch(t, cluster_name='test')
```

Still shows:
```
(task, pid=2152) /s1:
(task, pid=2152) total 4
(task, pid=2152) drwxrwxr-x 3 ubuntu ubuntu 4096 Nov 11 20:06 mypath
(task, pid=2152) 
(task, pid=2152) /s3:
(task, pid=2152) total 4
(task, pid=2152) drwxrwxr-x 3 ubuntu ubuntu 4096 Nov 11 20:06 mypath
```

Reply:
COPY mode supported now!